// @harmoniclabs/cardano-ledger-ts/dist/credentials/PaymentCredentials.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.PaymentCredentials = exports.ValidatorHash = void 0;
var cbor_1 = require("@harmoniclabs/cbor");
var plutus_data_1 = require("@harmoniclabs/plutus-data");
var Hash28_1 = require("../hashes/Hash28/Hash28.js");
var PubKeyHash_1 = require("./PubKeyHash.js");
var obj_utils_1 = require("@harmoniclabs/obj-utils");
var assert_1 = require("../utils/assert.js");
var ValidatorHash = /** @class */ (function (_super) {
    __extends(ValidatorHash, _super);
    function ValidatorHash() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return ValidatorHash;
}(Hash28_1.Hash28));
exports.ValidatorHash = ValidatorHash;
var PaymentCredentials = /** @class */ (function () {
    function PaymentCredentials(type, hash) {
        (0, assert_1.assert)(hash instanceof Hash28_1.Hash28, "can't construct 'PaymentCredentials'; hash must be instance of an 'Hash28'");
        (0, assert_1.assert)(type === "pubKey" || type === "script", "can't construct 'PaymentCredentials'; specified type is nor 'addres' nor 'script'");
        (0, obj_utils_1.defineReadOnlyProperty)(this, "type", type);
        (0, obj_utils_1.defineReadOnlyProperty)(this, "hash", type === "pubKey" ?
            (hash instanceof PubKeyHash_1.PubKeyHash ? hash : new PubKeyHash_1.PubKeyHash(hash.asBytes)) :
            (hash instanceof ValidatorHash ? hash : new ValidatorHash(hash.asBytes)));
    }
    PaymentCredentials.prototype.clone = function () {
        return new PaymentCredentials(this.type, this.hash.clone());
    };
    Object.defineProperty(PaymentCredentials, "fake", {
        get: function () {
            return new PaymentCredentials("pubKey", new Hash28_1.Hash28("ff".repeat(28)));
        },
        enumerable: false,
        configurable: true
    });
    PaymentCredentials.prototype.toData = function () {
        return new plutus_data_1.DataConstr(// PCredential
        this.type === "pubKey" ?
            0 : // PPubKeyCredential
            1, // PScriptCredential
        [
            // both bytestring alias as argument
            new plutus_data_1.DataB(this.hash.asBytes)
        ]);
    };
    PaymentCredentials.pubKey = function (hash) {
        return new PaymentCredentials("pubKey", hash instanceof PubKeyHash_1.PubKeyHash ?
            hash :
            new PubKeyHash_1.PubKeyHash(hash));
    };
    PaymentCredentials.script = function (hash) {
        return new PaymentCredentials("script", hash instanceof ValidatorHash ?
            hash :
            new ValidatorHash(hash));
    };
    PaymentCredentials.prototype.toCbor = function () {
        return cbor_1.Cbor.encode(this.toCborObj());
    };
    PaymentCredentials.prototype.toCborObj = function () {
        return new cbor_1.CborArray([
            new cbor_1.CborUInt(this.type === "pubKey" ? 0 : 1),
            this.hash.toCborObj()
        ]);
    };
    PaymentCredentials.fromCbor = function (cStr) {
        return PaymentCredentials.fromCborObj(cbor_1.Cbor.parse((0, cbor_1.forceCborString)(cStr)));
    };
    PaymentCredentials.fromCborObj = function (cObj) {
        if (!(cObj instanceof cbor_1.CborArray &&
            cObj.array[0] instanceof cbor_1.CborUInt))
            throw new Error("Invalid CBOR format for \"PaymentCredentials\"");
        return new PaymentCredentials(Number(cObj.array[0].num) === 0 ? "pubKey" : "script", Hash28_1.Hash28.fromCborObj(cObj.array[1]));
    };
    return PaymentCredentials;
}());
exports.PaymentCredentials = PaymentCredentials;
