// @harmoniclabs/plu-ts-onchain/dist/pluts/API/V1/ScriptContext/PTxInfo/PTxInfo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PTxInfo = void 0;
var pstruct_1 = require("../../../../PTypes/PStruct/pstruct.js");
var types_1 = require("../../../../type_system/types.js");
var PStakingCredential_1 = require("../../Address/PStakingCredential.js");
var PDCert_1 = require("../../PDCert/index.js");
var PPubKeyHash_1 = require("../../PubKey/PPubKeyHash.js");
var PDatumHash_1 = require("../../ScriptsHashes/PDatumHash.js");
var Time_1 = require("../../Time/index.js");
var PTxId_1 = require("../../Tx/PTxId.js");
var PTxInInfo_1 = require("../../Tx/PTxInInfo.js");
var PTxOut_1 = require("../../Tx/PTxOut.js");
var PValue_1 = require("../../Value/PValue/index.js");
exports.PTxInfo = (0, pstruct_1.pstruct)({
    PTxInfo: {
        inputs: (0, types_1.list)(PTxInInfo_1.PTxInInfo.type),
        outputs: (0, types_1.list)(PTxOut_1.PTxOut.type),
        fee: PValue_1.PValue.type,
        mint: PValue_1.PValue.type,
        dCertificates: (0, types_1.list)(PDCert_1.PDCert.type),
        withdrawals: (0, types_1.list)((0, types_1.pair)(PStakingCredential_1.PStakingCredential.type, types_1.int)),
        interval: Time_1.PPOSIXTimeRange.type,
        signatories: (0, types_1.list)(PPubKeyHash_1.PPubKeyHash.type),
        datums: (0, types_1.list)((0, types_1.pair)(PDatumHash_1.PDatumHash.type, types_1.data)),
        id: PTxId_1.PTxId.type
    }
});
