// @harmoniclabs/cardano-ledger-ts/dist/ledger/certs/CertStakeDeRegistration.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CertStakeDeRegistration = void 0;
var cbor_1 = require("@harmoniclabs/cbor");
var credentials_1 = require("../../credentials/index.js");
var roDescr_1 = require("../../utils/roDescr.js");
var CertificateType_1 = require("./CertificateType.js");
var plutus_data_1 = require("@harmoniclabs/plutus-data");
var maybeData_1 = require("../../utils/maybeData.js");
var defaultToDataVersion_1 = require("../../toData/defaultToDataVersion.js");
var getSubCborRef_1 = require("../../utils/getSubCborRef.js");
var CertStakeDeRegistration = /** @class */ (function () {
    function CertStakeDeRegistration(_a, subCborRef) {
        var stakeCredential = _a.stakeCredential;
        this.subCborRef = subCborRef;
        Object.defineProperties(this, {
            certType: __assign({ value: CertificateType_1.CertificateType.StakeDeRegistration }, roDescr_1.roDescr),
            stakeCredential: __assign({ value: stakeCredential }, roDescr_1.roDescr)
        });
    }
    CertStakeDeRegistration.prototype.toData = function (version) {
        version = (0, defaultToDataVersion_1.definitelyToDataVersion)(version);
        if (version === "v1" || version === "v2")
            return new plutus_data_1.DataConstr(1, [
                new plutus_data_1.DataConstr(0, // StakingCredential.StakingHash
                // credential
                [this.stakeCredential.toData(version)])
            ]);
        return new plutus_data_1.DataConstr(1, // PCertificate.StakeDeRegistration
        [
            this.stakeCredential.toData(version),
            (0, maybeData_1.nothingData)() // refound
        ]);
    };
    CertStakeDeRegistration.prototype.getRequiredSigners = function () {
        return [this.stakeCredential.hash.clone()];
    };
    CertStakeDeRegistration.prototype.toCbor = function () {
        if (this.subCborRef instanceof cbor_1.SubCborRef) {
            // TODO: validate cbor structure
            // we assume correctness here
            return new cbor_1.CborString(this.subCborRef.toBuffer());
        }
        return cbor_1.Cbor.encode(this.toCborObj());
    };
    CertStakeDeRegistration.prototype.toCborObj = function () {
        return new cbor_1.CborArray([
            new cbor_1.CborUInt(this.certType),
            this.stakeCredential.toCborObj()
        ]);
    };
    CertStakeDeRegistration.fromCborObj = function (cbor) {
        if (!(cbor instanceof cbor_1.CborArray &&
            cbor.array.length >= 2 &&
            cbor.array[0] instanceof cbor_1.CborUInt &&
            Number(cbor.array[0].num) === CertificateType_1.CertificateType.StakeDeRegistration))
            throw new Error("Invalid cbor for 'CertStakeDeRegistration'");
        return new CertStakeDeRegistration({
            stakeCredential: credentials_1.Credential.fromCborObj(cbor.array[1])
        }, (0, getSubCborRef_1.getSubCborRef)(cbor));
    };
    CertStakeDeRegistration.prototype.toJSON = function () { return this.toJson(); };
    CertStakeDeRegistration.prototype.toJson = function () {
        return {
            certType: (0, CertificateType_1.certTypeToString)(this.certType),
            stakeCredential: this.stakeCredential.toJson()
        };
    };
    return CertStakeDeRegistration;
}());
exports.CertStakeDeRegistration = CertStakeDeRegistration;
