// @harmoniclabs/cardano-ledger-ts/dist/ledger/certs/CertStakeRegistrationDeleg.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CertStakeRegistrationDeleg = void 0;
var cbor_1 = require("@harmoniclabs/cbor");
var credentials_1 = require("../../credentials/index.js");
var roDescr_1 = require("../../utils/roDescr.js");
var CertificateType_1 = require("./CertificateType.js");
var hashes_1 = require("../../hashes/index.js");
var ints_1 = require("../../utils/ints.js");
var plutus_data_1 = require("@harmoniclabs/plutus-data");
var getSubCborRef_1 = require("../../utils/getSubCborRef.js");
var CertStakeRegistrationDeleg = /** @class */ (function () {
    function CertStakeRegistrationDeleg(_a, subCborRef) {
        var stakeCredential = _a.stakeCredential, poolKeyHash = _a.poolKeyHash, coin = _a.coin;
        this.subCborRef = subCborRef;
        Object.defineProperties(this, {
            certType: __assign({ value: CertificateType_1.CertificateType.StakeRegistrationDeleg }, roDescr_1.roDescr),
            stakeCredential: __assign({ value: stakeCredential }, roDescr_1.roDescr),
            poolKeyHash: __assign({ value: new hashes_1.Hash28(poolKeyHash) }, roDescr_1.roDescr),
            coin: __assign({ value: (0, ints_1.forceBigUInt)(coin) }, roDescr_1.roDescr)
        });
    }
    CertStakeRegistrationDeleg.prototype.toData = function (version) {
        version = typeof version !== "string" ? "v3" : version;
        if (version !== "v3")
            throw new Error("Stake registration and delegation certificate only allowed in plutus v3");
        return new plutus_data_1.DataConstr(3, [
            this.stakeCredential.toData(version),
            this.poolKeyHash.toData(version),
            new plutus_data_1.DataI(this.coin)
        ]);
    };
    CertStakeRegistrationDeleg.prototype.getRequiredSigners = function () {
        return [this.stakeCredential.hash.clone()];
    };
    CertStakeRegistrationDeleg.prototype.toCbor = function () {
        if (this.subCborRef instanceof cbor_1.SubCborRef) {
            // TODO: validate cbor structure
            // we assume correctness here
            return new cbor_1.CborString(this.subCborRef.toBuffer());
        }
        return cbor_1.Cbor.encode(this.toCborObj());
    };
    CertStakeRegistrationDeleg.prototype.toCborObj = function () {
        return new cbor_1.CborArray([
            new cbor_1.CborUInt(this.certType),
            this.stakeCredential.toCborObj(),
            this.poolKeyHash.toCborObj(),
            new cbor_1.CborUInt(this.coin)
        ]);
    };
    CertStakeRegistrationDeleg.prototype.toJSON = function () { return this.toJson(); };
    CertStakeRegistrationDeleg.prototype.toJson = function () {
        return {
            certType: (0, CertificateType_1.certTypeToString)(this.certType),
            stakeCredential: this.stakeCredential.toJson(),
            poolKeyHash: this.poolKeyHash.toString(),
            coin: this.coin.toString()
        };
    };
    CertStakeRegistrationDeleg.fromCbor = function (cbor) {
        return CertStakeRegistrationDeleg.fromCborObj(cbor_1.Cbor.parse((0, cbor_1.forceCborString)(cbor)));
    };
    CertStakeRegistrationDeleg.fromCborObj = function (cbor) {
        if (!(cbor instanceof cbor_1.CborArray &&
            cbor.array.length >= 4 &&
            cbor.array[0] instanceof cbor_1.CborUInt &&
            Number(cbor.array[0].num) === CertificateType_1.CertificateType.StakeRegistrationDeleg &&
            cbor.array[3] instanceof cbor_1.CborUInt))
            throw new Error("invalid cbor for 'CertStakeRegistrationDeleg'");
        return new CertStakeRegistrationDeleg({
            stakeCredential: credentials_1.Credential.fromCborObj(cbor.array[1]),
            poolKeyHash: hashes_1.PoolKeyHash.fromCborObj(cbor.array[2]),
            coin: cbor.array[3].num
        }, (0, getSubCborRef_1.getSubCborRef)(cbor));
    };
    return CertStakeRegistrationDeleg;
}());
exports.CertStakeRegistrationDeleg = CertStakeRegistrationDeleg;
