// @harmoniclabs/cardano-ledger-ts/dist/ledger/certs/CertRegistrationDrep.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CertRegistrationDrep = void 0;
var cbor_1 = require("@harmoniclabs/cbor");
var credentials_1 = require("../../credentials/index.js");
var roDescr_1 = require("../../utils/roDescr.js");
var CertificateType_1 = require("./CertificateType.js");
var Anchor_1 = require("../../governance/Anchor.js");
var ints_1 = require("../../utils/ints.js");
var plutus_data_1 = require("@harmoniclabs/plutus-data");
var getSubCborRef_1 = require("../../utils/getSubCborRef.js");
var CertRegistrationDrep = /** @class */ (function () {
    function CertRegistrationDrep(_a, subCborRef) {
        var drepCredential = _a.drepCredential, coin = _a.coin, anchor = _a.anchor;
        this.subCborRef = subCborRef;
        Object.defineProperties(this, {
            certType: __assign({ value: CertificateType_1.CertificateType.RegistrationDrep }, roDescr_1.roDescr),
            drepCredential: __assign({ value: drepCredential }, roDescr_1.roDescr),
            coin: __assign({ value: (0, ints_1.forceBigUInt)(coin) }, roDescr_1.roDescr),
            anchor: __assign({ value: (0, Anchor_1.isIAnchor)(anchor) ? new Anchor_1.Anchor(anchor) : undefined }, roDescr_1.roDescr),
        });
    }
    CertRegistrationDrep.prototype.toData = function (version) {
        version = typeof version !== "string" ? "v3" : version;
        if (version !== "v3")
            throw new Error("DRep registration certificate only allowed in plutus v3");
        return new plutus_data_1.DataConstr(4, [
            this.drepCredential.toData(version),
            new plutus_data_1.DataI(this.coin)
        ]);
    };
    CertRegistrationDrep.prototype.getRequiredSigners = function () {
        return [this.drepCredential.hash.clone()];
    };
    CertRegistrationDrep.prototype.toCbor = function () {
        if (this.subCborRef instanceof cbor_1.SubCborRef) {
            // TODO: validate cbor structure
            // we assume correctness here
            return new cbor_1.CborString(this.subCborRef.toBuffer());
        }
        return cbor_1.Cbor.encode(this.toCborObj());
    };
    CertRegistrationDrep.prototype.toCborObj = function () {
        var _a, _b;
        return new cbor_1.CborArray([
            new cbor_1.CborUInt(this.certType),
            this.drepCredential.toCborObj(),
            new cbor_1.CborUInt(this.coin),
            (_b = (_a = this.anchor) === null || _a === void 0 ? void 0 : _a.toCborObj()) !== null && _b !== void 0 ? _b : new cbor_1.CborSimple(null)
        ]);
    };
    CertRegistrationDrep.fromCborObj = function (cbor) {
        if (!(cbor instanceof cbor_1.CborArray &&
            cbor.array.length >= 4 &&
            cbor.array[0] instanceof cbor_1.CborUInt &&
            Number(cbor.array[0].num) === CertificateType_1.CertificateType.RegistrationDrep &&
            cbor.array[2] instanceof cbor_1.CborUInt))
            throw new Error("Invalid cbor for 'CertRegistrationDrep'");
        return new CertRegistrationDrep({
            drepCredential: credentials_1.Credential.fromCborObj(cbor.array[1]),
            coin: cbor.array[2].num,
            anchor: cbor.array[3] instanceof cbor_1.CborSimple ? undefined : Anchor_1.Anchor.fromCborObj(cbor.array[3])
        }, (0, getSubCborRef_1.getSubCborRef)(cbor));
    };
    CertRegistrationDrep.prototype.toJSON = function () { return this.toJson(); };
    CertRegistrationDrep.prototype.toJson = function () {
        var _a, _b;
        return {
            certType: (0, CertificateType_1.certTypeToString)(this.certType),
            stakeCredential: this.drepCredential.toJson(),
            coin: this.coin.toString(),
            anchor: (_b = (_a = this.anchor) === null || _a === void 0 ? void 0 : _a.toJson()) !== null && _b !== void 0 ? _b : null
        };
    };
    return CertRegistrationDrep;
}());
exports.CertRegistrationDrep = CertRegistrationDrep;
