// @harmoniclabs/cardano-ledger-ts/dist/ledger/certs/CertUnRegistrationDrep.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CertUnRegistrationDrep = void 0;
var cbor_1 = require("@harmoniclabs/cbor");
var credentials_1 = require("../../credentials/index.js");
var roDescr_1 = require("../../utils/roDescr.js");
var CertificateType_1 = require("./CertificateType.js");
var ints_1 = require("../../utils/ints.js");
var plutus_data_1 = require("@harmoniclabs/plutus-data");
var getSubCborRef_1 = require("../../utils/getSubCborRef.js");
var CertUnRegistrationDrep = /** @class */ (function () {
    function CertUnRegistrationDrep(_a, subCborRef) {
        var drepCredential = _a.drepCredential, coin = _a.coin;
        this.subCborRef = subCborRef;
        Object.defineProperties(this, {
            certType: __assign({ value: CertificateType_1.CertificateType.UnRegistrationDrep }, roDescr_1.roDescr),
            drepCredential: __assign({ value: drepCredential }, roDescr_1.roDescr),
            coin: __assign({ value: (0, ints_1.forceBigUInt)(coin) }, roDescr_1.roDescr),
        });
    }
    CertUnRegistrationDrep.prototype.toData = function (version) {
        version = typeof version !== "string" ? "v3" : version;
        if (version !== "v3")
            throw new Error("DRep registration certificate only allowed in plutus v3");
        return new plutus_data_1.DataConstr(6, [
            this.drepCredential.toData(version),
            new plutus_data_1.DataI(this.coin)
        ]);
    };
    CertUnRegistrationDrep.prototype.getRequiredSigners = function () {
        return [this.drepCredential.hash.clone()];
    };
    CertUnRegistrationDrep.prototype.toCbor = function () {
        if (this.subCborRef instanceof cbor_1.SubCborRef) {
            // TODO: validate cbor structure
            // we assume correctness here
            return new cbor_1.CborString(this.subCborRef.toBuffer());
        }
        return cbor_1.Cbor.encode(this.toCborObj());
    };
    CertUnRegistrationDrep.prototype.toCborObj = function () {
        return new cbor_1.CborArray([
            new cbor_1.CborUInt(this.certType),
            this.drepCredential.toCborObj(),
            new cbor_1.CborUInt(this.coin)
        ]);
    };
    CertUnRegistrationDrep.fromCborObj = function (cbor) {
        if (!(cbor instanceof cbor_1.CborArray &&
            cbor.array.length >= 3 &&
            cbor.array[0] instanceof cbor_1.CborUInt &&
            Number(cbor.array[0].num) === CertificateType_1.CertificateType.UnRegistrationDrep &&
            cbor.array[2] instanceof cbor_1.CborUInt))
            throw new Error("Invalid cbor for 'CertUnRegistrationDrep'");
        return new CertUnRegistrationDrep({
            drepCredential: credentials_1.Credential.fromCborObj(cbor.array[1]),
            coin: cbor.array[2].num
        }, (0, getSubCborRef_1.getSubCborRef)(cbor));
    };
    CertUnRegistrationDrep.prototype.toJSON = function () { return this.toJson(); };
    CertUnRegistrationDrep.prototype.toJson = function () {
        return {
            certType: (0, CertificateType_1.certTypeToString)(this.certType),
            drepCredential: this.drepCredential.toJson(),
            coin: this.coin.toString()
        };
    };
    return CertUnRegistrationDrep;
}());
exports.CertUnRegistrationDrep = CertUnRegistrationDrep;
