// @harmoniclabs/cardano-ledger-ts/dist/governance/ProposalProcedure.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProposalProcedure = exports.isIProposalProcedure = void 0;
var cbor_1 = require("@harmoniclabs/cbor");
var ledger_1 = require("../ledger/index.js");
var Anchor_1 = require("./Anchor.js");
var GovActionLike_1 = require("./GovAction/GovActionLike.js");
var ints_1 = require("@harmoniclabs/cbor/dist/utils/ints");
var roDescr_1 = require("../utils/roDescr.js");
var obj_utils_1 = require("@harmoniclabs/obj-utils");
var ints_2 = require("../utils/ints.js");
var plutus_data_1 = require("@harmoniclabs/plutus-data");
function isIProposalProcedure(stuff) {
    return (0, obj_utils_1.isObject)(stuff) && ((0, ints_2.canBeUInteger)(stuff.deposit) &&
        stuff.rewardAccount instanceof ledger_1.StakeAddress &&
        (0, GovActionLike_1.isGovActionLike)(stuff.govAction) &&
        (0, Anchor_1.isIAnchor)(stuff.anchor));
}
exports.isIProposalProcedure = isIProposalProcedure;
var ProposalProcedure = /** @class */ (function () {
    function ProposalProcedure(_a, subCborRef) {
        var deposit = _a.deposit, rewardAccount = _a.rewardAccount, govAction = _a.govAction, anchor = _a.anchor;
        this.subCborRef = subCborRef;
        Object.defineProperties(this, {
            deposit: __assign({ value: (0, ints_1.forceBigUInt)(deposit) }, roDescr_1.roDescr),
            rewardAccount: __assign({ value: ledger_1.StakeAddress.fromString(rewardAccount.toString()) }, roDescr_1.roDescr),
            govAction: __assign({ value: (0, GovActionLike_1.toRealGovAction)(govAction) }, roDescr_1.roDescr),
            anchor: __assign({ value: new Anchor_1.Anchor(anchor) }, roDescr_1.roDescr)
        });
    }
    ProposalProcedure.prototype.toCbor = function () {
        if (this.subCborRef instanceof cbor_1.SubCborRef) {
            // TODO: validate cbor structure
            // we assume correctness here
            return new cbor_1.CborString(this.subCborRef.toBuffer());
        }
        return cbor_1.Cbor.encode(this.toCborObj());
    };
    ProposalProcedure.prototype.toCborObj = function () {
        return new cbor_1.CborArray([
            new cbor_1.CborUInt(this.deposit),
            this.rewardAccount.toCredential().toCborObj(),
            this.govAction.toCborObj(),
            this.anchor.toCborObj()
        ]);
    };
    ProposalProcedure.prototype.toData = function (v) {
        v = "v3"; // only supported so far
        return new plutus_data_1.DataConstr(0, [
            new plutus_data_1.DataI(this.deposit),
            this.rewardAccount.toCredential().toData(v),
            this.govAction.toData(v)
        ]);
    };
    return ProposalProcedure;
}());
exports.ProposalProcedure = ProposalProcedure;
