// @harmoniclabs/cardano-ledger-ts/dist/governance/GovAction/GovActionInitHardFork.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GovActionInitHardFork = exports.isIGovActionInitHardFork = void 0;
var cbor_1 = require("@harmoniclabs/cbor");
var tx_1 = require("../../tx/index.js");
var protocolVersion_1 = require("../../ledger/protocol/protocolVersion.js");
var GovActionType_1 = require("./GovActionType.js");
var roDescr_1 = require("../../utils/roDescr.js");
var obj_utils_1 = require("@harmoniclabs/obj-utils");
var plutus_data_1 = require("@harmoniclabs/plutus-data");
var maybeData_1 = require("../../utils/maybeData.js");
function isIGovActionInitHardFork(stuff) {
    return (0, obj_utils_1.isObject)(stuff) && ((stuff.govActionId === undefined || (0, tx_1.isITxOutRef)(stuff.govActionId)) &&
        (0, protocolVersion_1.isIProtocolVersion)(stuff.protocolVersion));
}
exports.isIGovActionInitHardFork = isIGovActionInitHardFork;
var GovActionInitHardFork = /** @class */ (function () {
    function GovActionInitHardFork(_a, subCborRef) {
        var govActionId = _a.govActionId, protocolVersion = _a.protocolVersion;
        this.subCborRef = subCborRef;
        Object.defineProperties(this, {
            govActionType: __assign({ value: GovActionType_1.GovActionType.InitHardFork }, roDescr_1.roDescr),
            govActionId: __assign({ value: govActionId ? new tx_1.TxOutRef(govActionId) : undefined }, roDescr_1.roDescr),
            protocolVersion: __assign({ value: (0, protocolVersion_1.protocolVersionAsObj)(protocolVersion) }, roDescr_1.roDescr)
        });
    }
    GovActionInitHardFork.prototype.toCbor = function () {
        if (this.subCborRef instanceof cbor_1.SubCborRef) {
            // TODO: validate cbor structure
            // we assume correctness here
            return new cbor_1.CborString(this.subCborRef.toBuffer());
        }
        return cbor_1.Cbor.encode(this.toCborObj());
    };
    GovActionInitHardFork.prototype.toCborObj = function () {
        var _a, _b;
        return new cbor_1.CborArray([
            new cbor_1.CborUInt(this.govActionType),
            (_b = (_a = this.govActionId) === null || _a === void 0 ? void 0 : _a.toCborObj()) !== null && _b !== void 0 ? _b : new cbor_1.CborSimple(null),
            (0, protocolVersion_1.protocolVersionToCborObj)(this.protocolVersion)
        ]);
    };
    GovActionInitHardFork.prototype.toData = function (v) {
        var _a;
        v = "v3"; // only one supported so far
        return new plutus_data_1.DataConstr(1, [
            (0, maybeData_1.maybeData)((_a = this.govActionId) === null || _a === void 0 ? void 0 : _a.toData(v)),
            new plutus_data_1.DataConstr(0, [
                new plutus_data_1.DataI(this.protocolVersion.major),
                new plutus_data_1.DataI(this.protocolVersion.minor)
            ])
        ]);
    };
    return GovActionInitHardFork;
}());
exports.GovActionInitHardFork = GovActionInitHardFork;
