// @harmoniclabs/cardano-ledger-ts/dist/governance/GovAction/GovActionTreasuryWithdrawals.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GovActionTreasuryWithdrawals = exports.isIGovActionTreasuryWithdrawals = void 0;
var cbor_1 = require("@harmoniclabs/cbor");
var hashes_1 = require("../../hashes/index.js");
var ledger_1 = require("../../ledger/index.js");
var GovActionType_1 = require("./GovActionType.js");
var roDescr_1 = require("../../utils/roDescr.js");
var obj_utils_1 = require("@harmoniclabs/obj-utils");
var plutus_data_1 = require("@harmoniclabs/plutus-data");
var maybeData_1 = require("../../utils/maybeData.js");
function isIGovActionTreasuryWithdrawals(stuff) {
    return (0, obj_utils_1.isObject)(stuff) && ((stuff.withdrawals instanceof ledger_1.TxWithdrawals || (0, ledger_1.isITxWithdrawals)(stuff.withdrawals)) &&
        (stuff.policyHash === undefined || (0, hashes_1.canBeHash28)(stuff.policyHash)));
}
exports.isIGovActionTreasuryWithdrawals = isIGovActionTreasuryWithdrawals;
var GovActionTreasuryWithdrawals = /** @class */ (function () {
    function GovActionTreasuryWithdrawals(_a, subCborRef) {
        var withdrawals = _a.withdrawals, policyHash = _a.policyHash;
        this.subCborRef = subCborRef;
        Object.defineProperties(this, {
            govActionType: __assign({ value: GovActionType_1.GovActionType.TreasuryWithdrawals }, roDescr_1.roDescr),
            withdrawals: __assign({ value: withdrawals instanceof ledger_1.TxWithdrawals ? withdrawals : new ledger_1.TxWithdrawals(withdrawals) }, roDescr_1.roDescr),
            policyHash: __assign({ value: (0, hashes_1.canBeHash28)(policyHash) ? new hashes_1.Hash28(policyHash) : undefined }, roDescr_1.roDescr)
        });
    }
    GovActionTreasuryWithdrawals.prototype.toCbor = function () {
        if (this.subCborRef instanceof cbor_1.SubCborRef) {
            // TODO: validate cbor structure
            // we assume correctness here
            return new cbor_1.CborString(this.subCborRef.toBuffer());
        }
        return cbor_1.Cbor.encode(this.toCborObj());
    };
    GovActionTreasuryWithdrawals.prototype.toCborObj = function () {
        var _a, _b;
        return new cbor_1.CborArray([
            new cbor_1.CborUInt(this.govActionType),
            this.withdrawals.toCborObj(),
            (_b = (_a = this.policyHash) === null || _a === void 0 ? void 0 : _a.toCborObj()) !== null && _b !== void 0 ? _b : new cbor_1.CborSimple(null)
        ]);
    };
    GovActionTreasuryWithdrawals.prototype.toData = function (v) {
        var _a;
        v = "v3"; // only one supported so far
        return new plutus_data_1.DataConstr(2, [
            this.withdrawals.toData(v),
            (0, maybeData_1.maybeData)((_a = this.policyHash) === null || _a === void 0 ? void 0 : _a.toData(v))
        ]);
    };
    return GovActionTreasuryWithdrawals;
}());
exports.GovActionTreasuryWithdrawals = GovActionTreasuryWithdrawals;
